core/src/main/scala/chisel3/package.scala

```diff
@@ -210,26 +211,140 @@ package object chisel3 {
+
+      // Align parts to their potential specifiers
+      val partsAndSpecifierSeq = partsAfterFirst.zip(args).map {
+        case (part, arg) => {
+
+          // Check if part starts with a format specifier (with % - disambiguate with literal % checking the next character if needed to be %)
+          // In the case of %f specifier there is a chance that we need more information - so capture till the 1st letter (a-zA-Z).
+          // Example cf"This is $val%2.2f here" - parts - Seq("This is ","%2.2f here") - the format specifier here is %2.2f.
+          val idx_of_fmt_str =
+            if (
+              !part.isEmpty() && part.charAt(0) == '%' && (part.size == 1 || (part.size >= 2 && part.charAt(1) != '%'))
+            ) part.indexWhere { _.isLetter }
+            else -1
+
+          val fmt = if (idx_of_fmt_str >= 0) Some(part.substring(0, idx_of_fmt_str + 1)) else None
+          val fmtArgs: Printable = arg match {
+            case d: Data => {
+              fmt match {
+                case Some("%n") => Name(d)
+                case Some("%N") => FullName(d)
+                case Some("%s") => d.toString
+                case Some(fForm) if d.isInstanceOf[Bits] => FirrtlFormat(fForm.substring(1,2),d)
+                case Some(x) => {
+                  val msg = s"Illegal format specifier '$x' for Chisel Data type!\n"
+                  throw new UnknownFormatConversionException(msg)
+                }
+                case None => d.toPrintable
+              }
+            }
+            case p: Printable => {
+              fmt match {
+                case Some("%s") => p.toString()
+                case Some(x) => {
+                  val msg = s"Illegal format specifier '$x' for Chisel Printable type!\n"
+                  throw new UnknownFormatConversionException(msg)
+                }
+                case None => p 
+              }
+            }
+
+            // Generic case - use String.format (for example %d,%2.2f etc on regular Scala types)
+            case t => {
+              PString(fmt.getOrElse("%s").format(t))
+            }
+          }
+
+          // Remove format specifier from parts string
+          val modP = part.zipWithIndex.collect { case (p, idx) if idx > idx_of_fmt_str => p }.mkString
+
+          (modP, Some(fmtArgs))
+        }
+      }
+      // Combine the 1st part with the rest of the modified (format specifier removed) parts
+      val combParts = parts(0) +: partsAndSpecifierSeq.map(_._1)
+
+      val pargsPables: Seq[Option[Printable]] = partsAndSpecifierSeq.map { _._2 }
       val seq = for { // append None because sc.parts.size == pargs.size + 1
-        (literal, arg) <- parts.zip(pargs :+ None)
-        optPable <- Seq(Some(PString(literal)), arg)
+        (literal, arg) <- combParts.zip(pargsPables :+ None)
+        optPable <- PercentSplitter(literal) ++ Seq(arg)
         pable <- optPable // Remove Option[_]
       } yield pable
       Printables(seq)
```

Comment:
If you do my suggested flatMap above, this just becomes
```scala
Printables(parts.head +: partsAndSpecifierSeq)
```
Should also rename `partsAndSpecifierSeq`.

Reply:
Refactored as suggested - except needed to ensure percentSplitter is called on parts.head as well.

Reply:
Renamed `partsAndSpecifierSeq` to `pables`